shared/src/shared.rs

```diff
@@ -276,6 +278,25 @@ impl<CI: ChainIndex> ChainProvider for Shared<CI> {
             .and_then(|ref root| self.store.get_transaction_meta(root, hash))
     }
 
+    // get block median time from this hash
+    fn block_median_time(&self, hash: &H256) -> Option<u64> {
+        let count = self.consensus.median_time_block_count();
+        let mut block_times = Vec::with_capacity(count);
+        let mut current_header = self.block_header(hash);
+        for _ in 0..count {
+            let header = match current_header {
+                Some(ref header) => header,
+                None => break,
+            };
+            block_times.push(header.timestamp());
+            if block_times.len() < count {
```

Comment:
Use `hash` as the iterator control variable, and call `block_header` at the beginning of the loop, you can eliminate this extra condition check.

Reply:
Because we release headers inside the loop, we need to call 10 times clone on `header.hash` to keep the outside `hash` variable.